### PR TITLE
Load optimizer separately for all current torch versions

### PR DESCRIPTION
**Pull request by snarayan21**
# What does this PR do?

Addresses a bug when loading optimizer state on resumption with torch >= 2.3.0, with `use_orig_params: False` and activation checkpointing. We currently do not load optimizer state separately when loading checkpoint for torch >=2.3.0, but according to torch's recommendation (see obscure comment [here](https://github.com/pytorch/pytorch/blob/f9fce110af428a600e597291a0ab80e43dc39e93/torch/distributed/checkpoint/optimizer.py#L215)), we should continue to load the optimizer state separately. This PR does that.

Filed an issue with PyTorch [here](https://github.com/pytorch/pytorch/issues/124546).

See Jira issue [here](https://databricks.atlassian.net/browse/GRT-2843).

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
@eracah is there any performance impact?

**Review by mvpatel2000 — COMMENTED** [review]
Can we please remove the torch gating checks instead and leave as is (separate loading)?

File: composer/utils/checkpoint.py [hunk 1]
```diff
@@ -1067,13 +1064,12 @@ def _save_checkpoint(
...
-        # To load optimizer states with 2.0 <= torch < 2.2.9 , the optimizer state must be at the top
-        # level of the state dict because the load_sharded_optimizer_state_dict function
+        # To load optimizer states with current torch versions, the optimizer state must be at the
+        # top level of the state dict because the load_sharded_optimizer_state_dict function
         # requires a top level state dict key for the optimizer.
         # See https://github.com/pytorch/pytorch/blob/v2.0.1/torch/distributed/checkpoint/optimizer.py#L271
         # for more info.
-        if version.parse(torch.__version__) < version.parse('2.2.9'):
-            state_dict['optimizers'] = state_dict['state'].pop('optimizers')
+        state_dict['optimizers'] = state_dict['state'].pop('optimizers')
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:1072]: Do we need this change where we change state dict? I really dont like optimizers at root level.

Comment (eracah) [reply composer/utils/checkpoint.py:1072]: wait what those comments are from 2 years ago

Comment (eracah) [reply composer/utils/checkpoint.py:1072]: what makes you think they still hold?

Comment (mvpatel2000) [reply composer/utils/checkpoint.py:1072]: Can we try it without?

Comment (snarayan21) [reply composer/utils/checkpoint.py:1072]: this bug makes me think it's still valid at least lol. I'll test

**Comment by snarayan21** [comment]
@mvpatel2000 after discussion offline with @eracah, this probably isn't the perferred fix here. It seems that loading the optimizer separately is fixing this issue sort of "on accident" and the real issue is why activation checkpointing is messing up the FQNs. I'm going to close this issue and try to find where the FQNs are getting messed up instead.

**Pull request closed without merging**